packages/rollup-core/src/app/data/producers/log-handlers.ts

```diff
@@ -0,0 +1,317 @@
+  )
+
+  let rollupTransaction: RollupTransaction
+  try {
+    // Skip the 4 bytes of MethodID
+    const calldata = remove0x(ethers.utils.hexDataSlice(tx.data, 4))
+    rollupTransaction = {
+      l1BlockNumber: tx.blockNumber,
+      l1Timestamp: tx.timestamp,
+      l1TxHash: l.transactionHash,
+      l1TxIndex: l.transactionIndex,
+      l1TxLogIndex: l.transactionLogIndex,
+      queueOrigin: QueueOrigin.SAFETY_QUEUE,
+      batchIndex: 0,
+      sender: add0x(calldata.substr(0, 40)),
+      target: add0x(calldata.substr(40, 40)),
+      // TODO Change nonce to a BigNumber so it can support 256 bits
+      nonce: new BigNumber(calldata.substr(80, 64), 'hex').toNumber(),
+      // TODO: Change gasLimit to a BigNumber so it can support 256 bits
+      gasLimit: new BigNumber(calldata.substr(144, 64), 'hex').toNumber(),
+      signature: add0x(calldata.substr(208, 130)),
+      calldata: add0x(calldata.substr(338)),
+    }
+  } catch (e) {
+    // This is, by definition, just an ill-formatted, and therefore invalid, tx.
+    log.debug(
+      `Error parsing calldata tx from CalldataTxEnqueued event. Calldata: ${tx.data}. Error: ${e.message}. Stack: ${e.stack}.`
+    )
+    return
+  }
+
+  await ds.insertL1RollupTransactions(l.transactionHash, [rollupTransaction])
+}
+
+/**
+ * Handles the L1ToL2BatchAppended event by parsing a RollupTransaction
+ * from the log event and storing it in the DB.
+ *
+ * @param ds The L1DataService to use for persistence.
+ * @param l The log event that was emitted.
+ * @param tx The transaction that emitted the event.
+ * @throws Error if there's an error with persistence.
+ */
+export const L1ToL2BatchAppendedLogHandler = async (
+  ds: L1DataService,
+  l: Log,
+  tx: TransactionResponse
+): Promise<void> => {
+  log.debug(
+    `L1ToL2BatchAppended event received at block ${tx.blockNumber}, tx ${l.transactionIndex}, log: ${l.transactionLogIndex}. TxHash: ${tx.hash}`
+  )
+  let batchNumber: number
+  try {
+    batchNumber = await ds.createNextL1ToL2Batch()
+  } catch (e) {
+    logError(
+      log,
+      `Error creating next L1ToL2Batch after receiving an event to do so!`,
+      e
+    )
+    throw e
+  }
+
+  if (!batchNumber) {
+    const msg = `Attempted to create L1 to L2 Batch upon receiving L1ToL2BatchAppended log, but no tx was available for batching!`
+    log.error(msg)
+    throw Error(msg)
+  } else {
+    log.debug(
+      `Successfully created L1 to L2 Batch! Batch number: ${batchNumber}`
+    )
+  }
+}
+
+/**
+ * Handles the SafetyQueueBatchAppended event by parsing a RollupTransaction
+ * from the transaction calldata and storing it in the DB.
+ *
+ * @param ds The L1DataService to use for persistence.
+ * @param l The log event that was emitted.
+ * @param tx The transaction that emitted the event.
+ * @throws Error if there's an error with persistence.
+ */
+export const SafetyQueueBatchAppendedLogHandler = async (
```

Comment:
It may not be called right after, but yes, it will be called after. Can't append to the canonical chain something that has not yet been queued. This is part of the motivation for the SQL DB. Since we know we had to have already received the "enqueued" event, parsed the tx, and stored it in the DB, when we get the "appended" event, it's just a DB update to say "Remember that thing that was queued? Change it from queued to part of the canonical chain."